src/Transports/MassTransit.KafkaIntegration/Contexts/ConsumerLockContext.cs

```diff
@@ -51,91 +67,57 @@ public void OnUnAssigned(IConsumer<TKey, TValue> consumer, IEnumerable<TopicPart
         {
             LogContext.SetCurrentIfNull(_hostConfiguration.ReceiveLogContext);
 
+            var tasks = new List<Task>();
             foreach (var partition in partitions)
             {
                 if (_data.TryRemove(partition.Partition, out var data))
-                    data.Close(partition);
+                    tasks.Add(data.Close(partition));
+            }
+
+            async Task Close()
+            {
+                await Task.WhenAll(tasks).ConfigureAwait(false);
             }
+
+            TaskUtil.Await(Close);
         }
 
 
         sealed class PartitionCheckpointData
         {
-            readonly IConsumer<TKey, TValue> _consumer;
-            readonly object _lock;
-            readonly int _maxCount;
-            readonly TopicPartition _partition;
-            readonly TimeSpan _timeout;
-            readonly Stopwatch _timer;
-            bool _commitIsRequired;
-            Offset _offset;
-            ushort _processed;
-
-            public PartitionCheckpointData(TopicPartition partition, IConsumer<TKey, TValue> consumer, TimeSpan timeout, ushort maxCount)
+            readonly ICheckpointer _checkpointer;
+            readonly ChannelExecutor _executor;
+            readonly PendingConfirmationCollection _pending;
+
+            public PartitionCheckpointData(TopicPartition partition, IConsumer<TKey, TValue> consumer, ReceiveSettings settings)
+            {
+                _pending = new PendingConfirmationCollection(partition);
+                _executor = new ChannelExecutor(1);
```

Comment:
Because I wanna executor per partition, and ConsumerContext is owning all partitions related to the topic

Reply:
Seems excessive? I don't know, why not just use topic+partition as a dictionary key?

Reply:
we can create additional one and share between partitions, but do not wanna use consumer one, really

Reply:
even though it is expensive, but partitions are running in separate thread, so I do not think it should be an issue, but it will provide really good encapsulation between them, otherwise one partition could lock another. And consumer is created only once for all partitions